src/dataprotection/azext_dataprotection/manual/custom.py

```diff
@@ -255,6 +255,132 @@ def dataprotection_backup_instance_list_from_resourcegraph(client, datasource_ty
     return response.data
 
 
+def dataprotection_backup_instance_update_msi_permissions(cmd, client, resource_group_name, datasource_type, vault_name, operation, permissions_scope, backup_instance=None, keyvault_id=None):
+    backup_vault = client.get(resource_group_name=resource_group_name,
+                              vault_name=vault_name)
+    principal_id = backup_vault.identity.principal_id
+
+    role_assignments_arr = []
+
+    if operation == 'Backup' and backup_instance is None:
+        raise CLIError("--backup-instance needs to be given when --operation is given as Backup")
+
+    if not keyvault_id and datasource_type == 'AzureDatabaseForPostgreSQL':
+        raise CLIError("--keyvault-id needs to be given when --datasource-type is AzureDatabaseForPostgreSQL")
+
+    from azure.cli.command_modules.role.custom import list_role_assignments, create_role_assignment
+
+    manifest = helper.load_manifest(datasource_type)
+    resource_map = {}
+
+    keyvault_client = None
+    keyvault = None
+    keyvault_subscription = None
+    keyvault_scope_arr = None
+    keyvault_name = None
+    keyvault_rg = None
+    if manifest['supportSecretStoreAuthentication']:
+        cmd.command_kwargs['operation_group'] = 'vaults'
+
+        from azure.cli.core.profiles import ResourceType
+        from azure.cli.command_modules.keyvault._client_factory import Clients
+        from azure.cli.core.commands.client_factory import get_mgmt_service_client
+
+        keyvault_scope_arr = keyvault_id.split("/")
+        keyvault_subscription = keyvault_scope_arr[2]
+        keyvault_name = keyvault_scope_arr[-1]
+        keyvault_rg = keyvault_scope_arr[4]
+
+        keyvault_client = getattr(get_mgmt_service_client(cmd.cli_ctx, ResourceType.MGMT_KEYVAULT, subscription_id=keyvault_subscription), Clients.vaults)
+
+        keyvault = keyvault_client.get(resource_group_name=keyvault_rg, vault_name=keyvault_name)
+
+        if keyvault.properties.public_network_access == 'Disabled':
+            raise CLIError("Keyvault needs to have public network access enabled")
+
+        keyvault_network_rules = keyvault.properties.network_acls
+        if keyvault_network_rules and keyvault_network_rules.default_action == 'Deny':
+            raise CLIError("Keyvault needs to have public network access enabled")
+
+        keyvault_permission_models = manifest['secretStorePermissions']
+        if keyvault.properties.enable_rbac_authorization:
+            keyvault_rbac_model = keyvault_permission_models['rbacModel']
+            role = keyvault_rbac_model['roleDefinitionName']
+
+            keyvault_assignment_scope = keyvault_id
+            if permissions_scope == 'ResourceGroup':
+                keyvault_assignment_scope = "/".join(keyvault_assignment_scope.split("/")[:5])
```

Comment:
Like the above comment you can use create helper functions like -

get_sub_id_from_arm_id, get_rg_id_from_arm_id and implement it there. Just call that function here so the code is reusable and clean.